lldb/tools/lldb-rpc/lldb-rpc-gen/client/RPCLibrarySourceEmitter.cpp

```diff
@@ -0,0 +1,542 @@
+    // variable to be marked `const`. That would prevent us from changing it
+    // during the decoding step.
+    QualType UnqualifiedReturnType = method.ReturnType.getUnqualifiedType();
+    ReturnValueStorageStream
+        << lldb_rpc_gen::ReplaceLLDBNamespaceWithRPCNamespace(
+               UnqualifiedReturnType.getAsString(method.Policy))
+        << " " << ReturnVariableName << " = {};";
+  }
+  EmitLine(ReturnValueStorage);
+}
+
+void RPCLibrarySourceEmitter::EmitConnectionSetup(const Method &method) {
+  // Methods know if they require a connection parameter. We need to figure out
+  // which scenario we're in.
+  bool ConnectionDerived = false;
+  if (!method.RequiresConnectionParameter()) {
+    // If we have an instance method that is not a constructor, we have a valid
+    // connection from `this` via `ObjectRefGetConnectionSP()`.
+    if (!method.IsCtor && method.IsInstance) {
+      EmitLine("// Deriving connection from this.");
+      EmitLine("rpc_common::ConnectionSP connection_sp = "
+               "ObjectRefGetConnectionSP();");
+      ConnectionDerived = true;
+    }
+
+    // Otherewise, we try to derive it from an existing parameter.
+    if (!ConnectionDerived)
+      for (const auto &Param : method.Params) {
+        if (lldb_rpc_gen::TypeIsSBClass(Param.Type)) {
+          EmitLine("// Deriving connection from SB class parameter.");
+          std::string ConnectionLine =
+              "rpc_common::ConnectionSP connection_sp = " + Param.Name;
+          if (Param.Type->isPointerType())
+            ConnectionLine += "->";
+          else
+            ConnectionLine += ".";
+          ConnectionLine += "ObjectRefGetConnectionSP();";
+          EmitLine(ConnectionLine);
+          ConnectionDerived = true;
+          break;
+        }
+      }
+  } else {
+    // This method requires a connection parameter. It will always be named
+    // "connection" and it will always come first in the parameter list.
+    EmitLine("// Using connection parameter.");
+    EmitLine(
+        "rpc_common::ConnectionSP connection_sp = connection.GetConnection();");
+    ConnectionDerived = true;
+  }
+
+  assert(ConnectionDerived &&
+         "Unable to determine where method should derive connection from");
+
+  // NOTE: By this point, we should have already emitted the storage for the
+  // return value.
+  std::string FailureReturnExpression;
+  if (method.ReturnType->isPointerType())
+    FailureReturnExpression = "nullptr";
+  else if (!method.ReturnType->isVoidType())
+    FailureReturnExpression = "__result";
+  EmitLine("if (!connection_sp) return " + FailureReturnExpression + ";");
+}
+
+void RPCLibrarySourceEmitter::EmitEncodeParameters(const Method &method) {
+  // Encode parameters.
+  if (method.IsInstance && !method.IsCtor)
+    EmitLine("RPCValueEncoder(send, "
+             "rpc_common::RPCPacket::ValueType::Argument, *this);");
+
+  for (auto Iter = method.Params.begin(); Iter != method.Params.end(); Iter++) {
```

Comment:
If Iter is not modified within the loop, use a range based for if you can make it work.

Reply:
Oh no it is modified.
```
      Iter++;
      StringListLine += Iter->Name + ");";
      Iter--;
```
Which is some, shall I say, "interesting" code :)

In this case you could use std::next(Iter)->Name rather than undoing the increment.

And you can't use a range based for because you need to peek forwards, but certainly you can clean up the handling of Iter a bit.